lib/gc/Transforms/Utils/ValueUtils.cpp

```diff
@@ -116,5 +116,34 @@ FailureOr<SmallVector<int64_t>> getStaticStrides(Value value) {
   return strides;
 }
 
+std::pair<Value, Value> getPtrAndOffset(OpBuilder &builder, Value operand) {
+  auto memrefType = dyn_cast<MemRefType>(operand.getType());
+  assert(memrefType && "Expect a memref value");
+
+  Location loc = operand.getDefiningOp()->getLoc();
+  OpBuilder::InsertionGuard guard(builder);
+  // Insert right after operand producer for better opt chances.
+  builder.setInsertionPointAfterValue(operand);
+
+  MemRefType baseMemrefType = MemRefType::get({}, memrefType.getElementType());
+  Type basePtrType = builder.getIndexType();
+  Type offsetType = builder.getIndexType();
+  SmallVector<Type> sizesTypes(memrefType.getRank(), offsetType);
+  SmallVector<Type> stridesTypes(memrefType.getRank(), offsetType);
+  auto meta = builder.create<memref::ExtractStridedMetadataOp>(
+      loc, baseMemrefType, offsetType, sizesTypes, stridesTypes, operand);
+  Value alignedPointerAsIndex =
+      builder.create<memref::ExtractAlignedPointerAsIndexOp>(loc, basePtrType,
+                                                             operand);
+  Value alignedPointerAsI64 = builder.create<arith::IndexCastOp>(
+      loc, builder.getIntegerType(64), alignedPointerAsIndex);
+  // TODO: non-POD will require an LLVMTypeConverter.
```

Comment:
POD?

Reply:
this is a util borrowed from tpp, POD here means 'Plain old data' which basically means simple C struct.